### PR TITLE
Alert Monitor Alert - Move Rest Client to Common

### PR DESCRIPTION
**Pull request by vincentdass**
- [x] Move Rest Client to Common

- [x] L0 Tests

- [x] E2E Tests

### CODE REVIEW

File: Tasks/AzureMonitorAlertsV0/npm-shrinkwrap.json [hunk 1]
```diff
@@ -4,31 +4,18 @@
...
   "dependencies": {
-    "azurestack-common": {
-      "version": "file:../../_build/Tasks/Common/azurestack-common-1.0.0.tgz",
+    "azure-arm-rest": {
+      "version": "file:../../_build/Tasks/Common/azure-arm-rest-1.0.2.tgz",
       "requires": {
         "q": "1.4.1",
         "typed-rest-client": "0.12.0",
-        "vsts-task-lib": "2.0.1-preview"
+        "vsts-task-lib": "2.0.5"
```
Comment (kmkumaran) [inline Tasks/AzureMonitorAlertsV0/npm-shrinkwrap.json]: Any new nested modules coming in part of this?

File: Tasks/AzureMonitorAlertsV0/azuremonitoralerts.ts [hunk 2]
```diff
@@ -1,13 +1,9 @@
 import * as tl from "vsts-task-lib/task";
 import * as path from "path";
+import { AzureRMEndpoint } from 'azure-arm-rest/azure-arm-endpoint';
```
Comment (kmkumaran) [inline Tasks/AzureMonitorAlertsV0/azuremonitoralerts.ts]: @asranja - Can you review this PR.

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+		let deferred: Q.Deferred<any> = Q.defer<any>();
+
+		let splittedResourceUri: string[] = resourceUri.split("/");
+		let resourceGroupName: string = splittedResourceUri[4];
+		let resourceType: string = splittedResourceUri[6] + "/" +  splittedResourceUri[7];
+		let resourceName: string = splittedResourceUri[8];
+
+		tl.debug(`Getting AzureRm resource details - '${resourceName}' in resource group '${resourceGroupName}'`);
+        
+        let resources: Resources = new Resources(this._azureEndpoint);
```
Comment (asranja) [inline Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: Indentation

File: Tasks/AzureMonitorAlertsV0/task.json [hunk 4]
```diff
@@ -13,7 +13,7 @@
 	"version": {
 		"Major": 0,
 		"Minor": 1,
```
Comment (asranja) [inline Tasks/AzureMonitorAlertsV0/task.json]: 0.2.0

File: Tasks/AzureMonitorAlertsV0/task.json [hunk 5]
```diff
@@ -130,6 +130,7 @@
 		"UpdatedRule": "Updated rule : '%s'",
 		"Couldnotfetchaccesstoken": "Could not fetch access token for Azure. Status Code: %s (%s) %s.",
 		"SPNExpiredCheck": "Check if the SPN is valid and not expired.",
-		"MSINotSupported": "Managed Service Identity(MSI) authentication is not supported for this task."
+		"FailedToGetApplicationInsightsResourceAlerts": "Failed to get Application Insights Alert rule: %s. Error: %s",
+        "FailedToUpdateApplicationInsightsResourceAlerts": "Failed to update Application Insights Alert rule '%s' Resource. Error: %s"
```
Comment (asranja) [inline Tasks/AzureMonitorAlertsV0/task.json]: Indentation

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+	}
+}
+
+
+
+export function getDeploymentUri(): string {
+	let buildUri = tl.getVariable("Build.BuildUri");
+	let releaseWebUrl = tl.getVariable("Release.ReleaseWebUrl");
+	let collectionUrl = tl.getVariable('System.TeamFoundationCollectionUri'); 
+    let teamProject = tl.getVariable('System.TeamProjectId');
```
Comment (asranja) [inline Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: indentation

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        let azureApplicationInsightsAlerts :AzureMonitorAlerts = new AzureMonitorAlerts(this._azureEndpoint, this._resourceGroupName);
+        for(let rule of alertRules) {
+            let requestBody: IAzureMetricAlertRequestBody = await this._getRequestBodyForAddingAlertRule(azureApplicationInsightsAlerts, this._resourceGroupName, resourceId, rule, notifyServiceOwners, notifyEmails);
+            await azureApplicationInsightsAlerts.update(rule.alertName, requestBody);
+        }
+    }
+
+
+    private async _getRequestBodyForAddingAlertRule(
+        azureApplicationInsightsAlerts :AzureMonitorAlerts,
```
Comment (asranja) [inline Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: rename azureApplicationInsightsAlerts.

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+		tl.debug(`Getting AzureRm resource details - '${resourceName}' in resource group '${resourceGroupName}'`);
+		let resources: Resources = new Resources(this._azureEndpoint);
+		let resourceValues = await resources.getResources(resourceType, resourceName);
+
+		return resourceValues[0];
+	}
+}
+
+
+
```
Comment (asranja) [inline Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: Too many empty lines.

File: Tasks/Common/azure-arm-rest/azure-arm-appinsigths-alerts.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        tl.debug(`Getting AzureRm alert rule - '${alertRuleName}' in resource group '${this._resourceGroupName}'`);
+
+        var httpRequest = new webClient.WebRequest();
+        httpRequest.method = 'GET';
+
+        httpRequest.uri = this._client.getRequestUri(`//subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/microsoft.insights/alertrules/{resourceName}`,
+        {
+            '{resourceGroupName}': this._resourceGroupName,
+            '{resourceName}': alertRuleName,
+        }, null, APIVersions.azure_arm_appinsights_alerts);
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/azure-arm-appinsigths-alerts.ts]: rename azure_arm_appinsights_alerts.

File: Tasks/Common/azure-arm-rest/constants.ts [hunk 10]
```diff
@@ -9,7 +9,8 @@ export const productionSlot: string = "production";
 export const mysqlApiVersion: string = '2017-12-01';
 
 export const APIVersions = {
-    azure_arm_appinsights: '2015-05-01'
+    azure_arm_appinsights: '2015-05-01',
+    azure_arm_appinsights_alerts: '2016-03-01'
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/constants.ts]: rename azure_arm_appinsights_alerts

File: Tasks/Common/azure-arm-rest/module.json [hunk 11]
```diff
@@ -135,6 +135,8 @@
...
         "ARG_RedirectResponseInvalidStatusCode": "The HTTP response code: '%s' is not a valid redirect status code.",
         "ARG_RedirectResponseLocationHeaderIsNull": "Location header is null for HTTP response with status code: %s",
-        "ASE_SSLIssueRecommendation": "To use a certificate in App Service, the certificate must be signed by a trusted certificate authority. If your web app gives you certificate validation errors, you're probably using a self-signed certificate and to resolve them you need to set a variable named VSTS_ARM_REST_IGNORE_SSL_ERRORS to the value true in the build or release definition"
+        "ASE_SSLIssueRecommendation": "To use a certificate in App Service, the certificate must be signed by a trusted certificate authority. If your web app gives you certificate validation errors, you're probably using a self-signed certificate and to resolve them you need to set a variable named VSTS_ARM_REST_IGNORE_SSL_ERRORS to the value true in the build or release definition",
+        "FailedToGetApplicationInsightsResourceAlerts": "Failed to get Application Insights Alert rule: %s. Error: %s",
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/module.json]: rename FailedToGetApplicationInsightsResourceAlerts.

File: Tasks/Common/azure-arm-rest/module.json [hunk 12]
```diff
@@ -135,6 +135,8 @@
...
+        "FailedToUpdateApplicationInsightsResourceAlerts": "Failed to update Application Insights Alert rule '%s' Resource. Error: %s"
```
Comment (asranja) [inline Tasks/Common/azure-arm-rest/module.json]: Rename FailedToUpdateApplicationInsightsResourceAlerts

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+        this._resourceGroupName = resourceGroupName;
+        this._resourceType = resourceType;
+        this._resourceName = resourceName;
+    }
+
+    public async addOrUpdateAlertRules(alertRules: IAzureMetricAlertRule[], notifyServiceOwners: boolean, notifyEmails: string) {
+
+        let resourceId: string = `/subscriptions/${this._azureEndpoint.subscriptionID}/resourceGroups/${this._resourceGroupName}/providers/${this._resourceType}/${this._resourceName}`;
+        
+        let azureApplicationInsightsAlerts :AzureMonitorAlerts = new AzureMonitorAlerts(this._azureEndpoint, this._resourceGroupName);
```
[the remark this answers is not in the document]
Comment (vincentdass) [reply Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: done

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts (hunk 7 above)
Comment (vincentdass) [reply Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: done

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+			let existingAlertRuleTargetResourceUri: string = existingAlertRule["properties"] && existingAlertRule["properties"].condition
+				&& existingAlertRule["properties"].condition.dataSource ? existingAlertRule["properties"].condition.dataSource.resourceUri: "";
+			if(existingAlertRuleTargetResourceUri && existingAlertRuleTargetResourceUri.toLowerCase() !== resourceUri.toLowerCase()) {
+				return Q.reject(tl.loc("AlertRuleTargetResourceIdMismatchError", rule.alertName, existingAlertRuleTargetResourceUri));
+			}
+
+			console.log(tl.loc("AlertRuleExists", rule.alertName, resourceGroupName));
+		}
+		catch (error) {
+            console.log(error);
```
[the remark this answers is not in the document]
Comment (vincentdass) [reply Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: done

File: Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts (hunk 8 above)
Comment (vincentdass) [reply Tasks/AzureMonitorAlertsV0/operations/AzureMonitorAlertsUtility.ts]: removed

File: Tasks/Common/azure-arm-rest/azure-arm-appinsigths-alerts.ts (hunk 9 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/azure-arm-appinsigths-alerts.ts]: done

File: Tasks/Common/azure-arm-rest/constants.ts (hunk 10 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/constants.ts]: done

File: Tasks/Common/azure-arm-rest/module.json (hunk 11 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/module.json]: done

File: Tasks/Common/azure-arm-rest/module.json (hunk 12 above)
Comment (vincentdass) [reply Tasks/Common/azure-arm-rest/module.json]: done

**Pull request merged**